File: tests/test_coverage_backlog.py
import json
import tempfile
import unittest
from pathlib import Path

from agents.coverage_backlog import (
    claim_batch,
    complete_batch,
    ensure_ledger,
    fail_batch,
    main,
    shard_items,
    unresolved_items,
)


class CoverageBacklogTests(unittest.TestCase):
    def state(self):
        return {
            "schema": "auto-dig-prolog-state.v1",
            "last_issue": None,
            "last_priority": None,
            "last_branch": None,
            "last_run_id": None,
            "last_success_at": None,
        }

    def ledger(self):
        state = self.state()
        _, ledger = ensure_ledger(
            state,
            corpus="anarchist-violence",
            seed_kind="us-geographic-npa",
            authority="NANPA/FCC",
            shard="B",
            shard_count=3,
            shard_index=1,
        )
        return state, ledger

    def test_sharding_uses_only_authoritative_input_items(self):
        valid = [202, 205, 208, 212, 213, 214, 215]
        self.assertEqual(shard_items(valid, 3, 1), [202, 205, 208, 214])
        self.assertNotIn(211, shard_items(valid, 3, 1))

    def test_claim_is_numeric_bounded_and_no_repeat(self):
        _, ledger = self.ledger()
        valid = [202, 205, 208, 214, 217]
        ledger["completed"] = [202]
        first = claim_batch(ledger, valid, 2)
        second = claim_batch(ledger, valid, 2)
        self.assertEqual(first, [205, 208])
        self.assertEqual(second, [214, 217])
        self.assertEqual(ledger["in_progress"], [205, 208, 214, 217])

    def test_completed_items_never_reenter_unresolved(self):
        _, ledger = self.ledger()
        valid = [202, 205, 208]
        self.assertEqual(claim_batch(ledger, valid, 2), [202, 205])
        complete_batch(ledger, [202, 205])
        self.assertEqual(unresolved_items(ledger, valid), [208])
        self.assertEqual(ledger["completed"], [202, 205])

    def test_failed_items_become_retryable_and_reenter_queue(self):
        _, ledger = self.ledger()
        valid = [202, 205]
        self.assertEqual(claim_batch(ledger, valid, 1), [202])
        fail_batch(ledger, [202])
        self.assertEqual(ledger["failed_retryable"], [202])
        self.assertEqual(unresolved_items(ledger, valid), [202, 205])
        self.assertEqual(claim_batch(ledger, valid, 1), [202])
        self.assertEqual(ledger["failed_retryable"], [])

    def test_new_authoritative_item_appears_without_resetting_state(self):
        _, ledger = self.ledger()
        ledger["completed"] = [202, 205, 208]
        self.assertEqual(unresolved_items(ledger, [202, 205, 208]), [])
        self.assertEqual(unresolved_items(ledger, [202, 205, 208, 214]), [214])

    def test_ledger_identity_cannot_drift(self):
        state, _ = self.ledger()
        with self.assertRaises(ValueError):
            ensure_ledger(
                state,
                corpus="anarchist-violence",
                seed_kind="us-geographic-npa",
                authority="NANPA/FCC",
                shard="B",
                shard_count=3,
                shard_index=2,
            )

    def test_cli_persists_claim_atomically(self):
        with tempfile.TemporaryDirectory() as directory:
            root = Path(directory)
            state_path = root / "state.json"
            valid_path = root / "valid.json"
            state_path.write_text(json.dumps(self.state()), encoding="utf-8")
            valid_path.write_text(json.dumps([202, 205, 208, 214]), encoding="utf-8")
            rc = main(
                [
                    "--state",
                    str(state_path),
                    "--valid-items",
                    str(valid_path),
                    "--corpus",
                    "anarchist-violence",
                    "--seed-kind",
                    "us-geographic-npa",
                    "--authority",
                    "NANPA/FCC",
                    "--shard",
                    "B",
                    "--shard-count",
                    "3",
                    "--shard-index",
                    "1",
                    "--batch-size",
                    "2",
                ]
            )
            self.assertEqual(rc, 0)
            stored = json.loads(state_path.read_text(encoding="utf-8"))
            ledger = stored["coverage"]["anarchist-violence/us-geographic-npa/B"]
            self.assertEqual(ledger["in_progress"], [202, 205])


if __name__ == "__main__":
    unittest.main()

File: agents/test_coverage_backlog.pl
:- begin_tests(coverage_backlog).

:- use_module('./coverage_backlog').

base_state(_{schema:"auto-dig-prolog-state.v1",
             last_issue:null,
             last_priority:null,
             last_branch:null,
             last_run_id:null,
             last_success_at:null}).

ledger(State, Ledger) :-
    ensure_ledger(State, "anarchist-violence", "us-geographic-npa", "NANPA",
                  "B", 3, 1, State1, Key),
    Ledger = State1.coverage.get(Key).

test(shard_uses_only_authoritative_items) :-
    shard_items([202,203,204,205,208,999], 3, 1, Items),
    assertion(Items == [202,205,208]).

test(completed_items_do_not_repeat) :-
    base_state(State), ledger(State, L0),
    claim_batch(L0, [202,205,208,211], 2, L1-B1),
    assertion(B1 == [202,205]),
    complete_batch(L1, B1, L2),
    claim_batch(L2, [202,205,208,211], 3, _L3-B2),
    assertion(B2 == [208]).

test(failed_items_become_retryable) :-
    base_state(State), ledger(State, L0),
    claim_batch(L0, [202,205], 1, L1-B1),
    fail_batch(L1, B1, L2),
    assertion(L2.failed_retryable == [202]),
    claim_batch(L2, [202,205], 1, L3-B2),
    assertion(B2 == [202]),
    assertion(L3.failed_retryable == []).

test(new_authoritative_item_appears_later) :-
    base_state(State), ledger(State, L0),
    claim_batch(L0, [202], 1, L1-B1),
    complete_batch(L1, B1, L2),
    claim_batch(L2, [202,205], 2, _L3-B2),
    assertion(B2 == [205]).

test(identity_drift_fails,
     [throws(error(permission_error(change, coverage_ledger_identity, authority), _))]) :-
    base_state(State),
    ensure_ledger(State, "anarchist-violence", "us-geographic-npa", "NANPA",
                  "B", 3, 1, State1, _),
    ensure_ledger(State1, "anarchist-violence", "us-geographic-npa", "FCC",
                  "B", 3, 1, _, _).

:- end_tests(coverage_backlog).
